File: Data Structures & Algorithms/target-sum/submission-1.cpp
class Solution {
public:
    int solve(vector<int>& nums, int target, int i, map<pair<int, int>, int>& dp) {
        if(target == 0 && i == nums.size()) return 1;
        if(i>=nums.size()) return 0;
        if(dp.contains({i, target})) return dp[{i, target}];
        return dp[{i, target}] = solve(nums, target-nums[i], i+1, dp) + solve(nums, target+nums[i], i+1, dp);
    }

    int findTargetSumWays(vector<int>& nums, int target) {
        map<pair<int, int>, int> dp;
        return solve(nums, target, 0, dp);
    }
};

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        sum = 0
        for num in nums:
            sum += num
        if sum%2 != 0:
            return False
        target = int(sum/2)
        n = len(nums)
        dp = [[-1] * (target + 1) for _ in range(n+1)]
        def solve(i: int, sum: int) -> bool:
            if sum == 0:
                return True
            if sum < 0 or i==len(nums):
                return False
            if dp[i][sum] != -1:
                return dp[i][sum]
            dp[i][sum] = solve(i + 1, sum - nums[i]) or solve(i+1, sum)
            return dp[i][sum]
        return solve(0, target)
